### ISSUE TITLE
Performance/HashEachMethods does not provide a net performance benefit

### ISSUE THREAD

**Issue by urbanautomaton** [bbatsov/rubocop#5589]
## Background

The Performance/HashEachMethods cop was introduced in af84fb2b4, in PR https://github.com/bbatsov/rubocop/pull/2578.

Since its introduction, there have been a large number of issues raised related to false positives (https://github.com/bbatsov/rubocop/issues/5328, https://github.com/bbatsov/rubocop/issues/5028, https://github.com/bbatsov/rubocop/issues/4931, https://github.com/bbatsov/rubocop/pull/4883, https://github.com/bbatsov/rubocop/issues/4881, https://github.com/bbatsov/rubocop/issues/4872, https://github.com/bbatsov/rubocop/issues/4777, https://github.com/bbatsov/rubocop/issues/4774, https://github.com/bbatsov/rubocop/pull/4754, https://github.com/bbatsov/rubocop/issues/4732).

While some of these have been addressed, the fundamental problem that the cop can't know what type of object is receiving the `#each` message means that most of these false positives will never be eliminated. The cop will match any invocation of `#each` yielding two arguments of which one is ignored: for example, iterating over an array of pairs.

As a workaround, the cop suggests wrapping destructuring block arguments in parens:

```ruby
# violates the cop
[[1, 2], [3, 4]].each do |k, _|
  p k
end

# does not violate the cop
[[1, 2], [3, 4]].each do |(k, _)|
  p k
end
```

However, measurement suggests the cop's suggested changes do not reliably result in performance improvements, and that the suggested workaround for its false positives usually *does* have a negative performance impact.

## Expected behavior

The `Performance/HashEachMethods` cop's suggested changes should improve code performance.

## Actual behavior

The cop's suggested changes do not significantly improve performance in recent rubies, and the suggested workaround for its false positives significantly harms performance.

## Steps to reproduce the problem

I [ran a benchmark](https://gist.github.com/urbanautomaton/7051549d5873964a878d5af3b8a7da1c) comparing `Hash#each` with `Hash#each_key` and `Hash#each_value`, and another comparing `Array#each` on an array of pairs, using implicit and explicit destructuring args.

The full benchmark code and results can be seen [here](https://gist.github.com/urbanautomaton/7051549d5873964a878d5af3b8a7da1c), and are summarised below (for simplicity's sake I've looked only at key-only access, but the results for value-only are broadly similar).

Hash comparison (`Hash#each` = 1.0, larger number implies faster, measurements with the margin of error in _italics_):

| Ruby: | 2.2 | 2.3 | 2.4 | 2.5 |
| --- | --- | --- | --- | --- |
| `Hash#each` | 1.0 | 1.0 | 1.0 | 1.0 |
| `Hash#each_key` | 1.16 | 1.15 | _1.02_ | _0.99_ |

Array comparison (`|k, _|` = 1.0, larger number implies faster, measurements within the margin of error in _italics_):

| Ruby: | 2.2 | 2.3 | 2.4 | 2.5 |
| --- | --- | --- | --- | --- |
| `Array#each` with `k, _` | 1.0 | 1.0 | 1.0 | 1.0 |
| `Array#each` with `(k, _)` | _0.95_ | 0.87 | 0.89 | 0.90 |

The small performance benefit to `Hash#each_key` that used to exist appears to have been eliminated in recent rubies, while a performance hit of similar magnitude remains when using parens for destructuring `Array#each` block arguments.

## Conclusion

I think we should remove this cop, as:

* it no longer provides performance benefits in recent ruby versions,
* it does cost non-trivial end user time working around false positives, and
* it suggests a workaround for its false positives that is harmful to performance.

I'd be happy to write a PR for this if it's agreed we should do so.

## RuboCop version

```
$ rubocop -V
0.52.1 (using Parser 2.4.0.2, running on ruby 2.4.3 x86_64-darwin17)
```

**Comment by rrosenblum** [comment]
Based on the performance numbers, I agree, this should be removed.
